mappings/muse.py

```diff
@@ -1,100 +1,170 @@
-        self.record.source = 'muse'
-        self.record.source_id = list(self.record.identifiers[0].split('|'))[0]
+        return new_record
+
+    def _get_formatted_field(self, record, field, subfields, string_format):
+        field_data = record.get_fields(field)
+        formatted_data = []
+        subfield_data = []
+
+        for data in field_data:
+            if data.is_control_field():
+                subfield_data.append([getattr(data, 'data')])
+            else:
+                subfield_data = [data.get_subfields(subfield) if data.get_subfields(subfield) else [''] for subfield in subfields]
+            
+            if subfield_data == [['']]:
+                continue
 
-        # Take first title as they are in order of preference
-        self.record.title = self.record.title[0]
+            for values in zip(*subfield_data):
+                formatted_string = string_format.format(*values)
+                formatted_data.append(utils.clean_formatted_string(formatted_string))
 
-        self.record.identifiers = [self.cleanup_identifier(id) for id in self.record.identifiers]
+        return formatted_data
+    
+    def _get_identifiers(self, record):
+        all_identifiers = []
         
-        self.record.languages = [
+        all_identifiers.extend(self._get_formatted_field(record, '001', ['0'], '{}|muse'))
+        all_identifiers.extend(self._get_formatted_field(record, '010', ['z'], '{}|lccn'))
+        all_identifiers.extend(self._get_formatted_field(record, '020', ['a', 'z'], '{}{}|isbn'))
+        all_identifiers.extend(self._get_formatted_field(record, '022', ['a'], '{}|issn'))
+        all_identifiers.extend(self._get_formatted_field(record, '035', ['a'], '{}|oclc'))
+        all_identifiers.extend(self._get_formatted_field(record, '040', ['a'], '{}|muse'))
+
+        return [self._cleanup_identifier(id) for id in all_identifiers]
+    
+    def _get_authors(self, record):
+        all_authors = []
+
+        all_authors.extend(self._get_formatted_field(record, '100', ['a', 'b', 'c', 'd'], '{} {} {} {}|||true'))
+        all_authors.extend(self._get_formatted_field(record, '110', ['a', 'b', 'c', 'n', 'd'], '{} {} {} {} {}|||true'))
+
+        return all_authors
+    
+    def _get_title(self, record):
+        title_245 = self._get_formatted_field(record, '245', ['a', 'b'], '{} {}')
+        title_130 = self._get_formatted_field(record, '130', ['a'], '{}')
+
+        all_titles = list(set(title_245 + title_130))
+
+        return all_titles[0]
+    
+    def _get_publishers(self, record):
+        publishers = self._get_formatted_field(record, '264', ['b'], '{}||')
+
+        if len(publishers) < 1:
+            publishers.append(DEFAULT_PUBLISHER)
+
+        return publishers
+    
+    def _get_dates(self, record):
+        dates = self._get_formatted_field(record, '264', ['c'], '{}|publication_date')
+
+        if len(dates) < 1:
+            publication_date = self.source['008'].data[11:15]
```

Comment:
We might want to use the record here instead of the source. But could this come back None?

Reply:
Yes, it can be none, updated to handle this.